pkg/treeoperator/controller/virtualcluster_join_controller.go

```diff
@@ -1,16 +1,452 @@
+		}
+	} else {
+		cluster.ResourceVersion = old.GetResourceVersion()
+		update, err := kosmosClient.KosmosV1alpha1().Clusters().Update(context.TODO(), &cluster, metav1.UpdateOptions{})
+		if err != nil {
+			return fmt.Errorf("update cluster %s failed: %v", clusterName, err)
+		} else {
+			klog.Infof("Cluster %s hase been updated.", clusterName)
+		}
+		if !update.DeletionTimestamp.IsZero() {
+			return fmt.Errorf("cluster %s is deleteting, need requeue", clusterName)
+		}
+	}
+	klog.Infof("Cluster %s has been created.", clusterName)
+	return nil
+}
+
+func (c *VirtualClusterJoinController) AddFinalizer(ctx context.Context, vc *v1alpha1.VirtualCluster) error {
+	vcNew := vc.DeepCopy()
+	if controllerutil.AddFinalizer(vcNew, constants.VirtualClusterSFinalizerName) {
+		err := c.Update(ctx, vcNew)
+		if err != nil {
+			return fmt.Errorf("add finalizer error for virtualcluster %s: %v", vc.Name, err)
+		}
+	}
+	klog.Infof("add finalizer for virtualcluster %s", vc.Name)
+	return nil
+}
+
+func (c *VirtualClusterJoinController) RemoveFinalizer(ctx context.Context, vc *v1alpha1.VirtualCluster) error {
+	vcNew := vc.DeepCopy()
+	if controllerutil.ContainsFinalizer(vcNew, constants.VirtualClusterSFinalizerName) {
+		controllerutil.RemoveFinalizer(vcNew, constants.VirtualClusterSFinalizerName)
+		err := c.Update(ctx, vcNew)
+		if err != nil {
+			return fmt.Errorf("remove finalizer error for virtualcluster %s: %v", vc.Name, err)
+		}
+	}
+	klog.Infof("remove finalizer for virtualcluster %s", vc.Name)
+	return nil
+}
+
+func (c *VirtualClusterJoinController) InstallClustertree(ctx context.Context, request reconcile.Request, vc *v1alpha1.VirtualCluster) error {
+	klog.Infof("Start creating kosmos-clustertree in namespace %s", request.Namespace)
+	defer klog.Infof("Finish creating kosmos-clustertree in namespace %s", request.Namespace)
+	if err := c.DeployKosmos(ctx, request, vc); err != nil {
+		return fmt.Errorf("deploy kosmos error: %v", err)
+	}
+	if err := c.CreateCluster(ctx, request, vc); err != nil {
+		return fmt.Errorf("create cluster error: %v", err)
+	}
+	if err := c.AddFinalizer(ctx, vc); err != nil {
+		return fmt.Errorf("add finalizer error: %v", err)
+	}
 	return nil
 }
+
+func (c *VirtualClusterJoinController) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	klog.V(4).Infof("============ %s starts to reconcile %s ============", constants.JoinControllerName, request.Name)
+	defer klog.V(4).Infof("============ %s reconcile finish %s ============", constants.JoinControllerName, request.Name)
+	var vc v1alpha1.VirtualCluster
+
+	if err := c.Get(ctx, request.NamespacedName, &vc); err != nil {
+		if errors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		klog.Errorf("get %s error: %v", request.NamespacedName, err)
+		return reconcile.Result{RequeueAfter: utils.DefaultRequeueTime}, nil
+	}
+	if vc.DeletionTimestamp.IsZero() {
```

Comment:
It is wrong to install services only based on whether vs is deleted or not.

Reply:
fix it